File: tutorials/mpi/examples/mpi_heat2D.f
C ****************************************************************************
C FILE: mpi_heat2D.f
C OTHER FILES: draw_heatf.c mpi_heat2D.h
C DESCRIPTION:  
C   HEAT2D Example - Parallelized Fortran Version
C   This example is based on a simplified two-dimensional heat
C   equation domain decomposition.  The initial temperature is computed to be
C   high in the middle of the domain and zero at the boundaries.  The
C   boundaries are held at zero throughout the simulation.  During the
C   time-stepping, an array containing two domains is used; these domains
C   alternate between old data and new data.
C 
C   In this parallelized version, the grid is decomposed by the master
C   process and then distributed by cols to workers processes.  At each
C   time step, workers processes must exchange border data with neighbors,
C   because a grid point's current temperature depends upon it's previous
C   time step value plus the values of the neighboring grid points.  Upon
C   completion of all time steps, the worker processes return their results
C   to the master process.
C 
C   Two data files are produced: an initial data set and a final data set.
C   An X graphic of these two states displays after all calculations have
C   completed.
C 
C AUTHOR: Blaise Barney - adapted from D. Turner's serial C version. Converted
C   to MPI: George L. Gusciora (1/95)
C LAST REVISED: 06/12/13 Blaise Barney
C ****************************************************************************
C Explanation of constants and variables
C   NXPROB                         =  x dimension of problem grid 
C   NYPROB                         =  y dimension of problem grid
C   STEPS                          =  number of time steps
C   MAXWORKER                      =  maximum number of workers tasks
C   MINWORKER                      =  minimum number of workers tasks
C   BEGIN, LTAG, RTAG, DONE        =  message tags
C   NONE                           =  indicates no neighbor
C   CX, CY                         =  used in heat equation
C   u                              =  array for grids
C   taskid,MASTER                  =  taskids 
C   numworkers                     =  number of workers processes
C   numtasks                       =  number of tasks
C   avecol,cols,offset,extra       =  for sending cols of data
C   dest, source                   =  to - from for message send-receive
C   left,right                     =  neighbor tasks
C   msgtype                        =  for message types
C   rc,start,end                   =  misc
C   i,ix,iy,iz,it                  =  loop variables
C ---------------------------------------------------------------------------

      program heat2D
      include 'mpif.h'
      include 'mpi_heat2D.h'

C     Routine for creating the X graph of the wave
      external draw_heat

      integer STEPS,MAXWORKER,MINWORKER,BEGIN,LTAG,RTAG,DONE,
     .        NONE,MASTER
      parameter(STEPS=50)
      parameter(MAXWORKER=8)
      parameter(MINWORKER=3)
      parameter(BEGIN=1)
      parameter(LTAG=2)
      parameter(RTAG=3)
      parameter(DONE=4)
      parameter(NONE=0)
      parameter(MASTER=0)
      integer taskid,numtasks,numworkers,avecol,cols,offset,extra,
     &        dest,source,left,right,msgtype,
     &        rc,start,end,i,ix,iy,iz,it,ierr
      integer status(MPI_STATUS_SIZE)

C     First, find out my taskid and how many tasks are running */
      call MPI_INIT( ierr )
      call MPI_COMM_RANK( MPI_COMM_WORLD, taskid, ierr )
      call MPI_COMM_SIZE( MPI_COMM_WORLD, numtasks, ierr )
      numworkers = numtasks-1

      if (taskid .eq. MASTER) then
C     ****************************** master code *****************************
C     Check if numworkers is within range - quit if not
      if ((numworkers.lt.MINWORKER).or.(numworkers.gt.MAXWORKER)) then
        print *,'MP_PROCS needs to be between', MINWORKER+1,'and',
     .           MAXWORKER+1, 'for this exercise'
        print *,'Quitting...'
        call MPI_ABORT(MPI_COMM_WORLD, rc, ierr)
      end if

      print *, 'Starting mpi_heat2D with',numworkers,'worker tasks.'

C     Initialize grid 
      print *,'Grid size: X=',NXPROB,' Y=',NYPROB,' Time steps=',STEPS
      print *,'Initializing grid and writing initial.dat file...'
      call inidat
      call prtdat(1)

C     Distribute work to workers.  Must first figure out how many cols to
C     send and what to do with extra cols.  
      avecol=NYPROB/numworkers
      extra=mod(NYPROB,numworkers)
      offset=1
      do i=1, numworkers
        if (i .le. extra) then
          cols=avecol+1
        else
          cols=avecol 
        end if
C       Tell each worker which other workers are its neighbors, since
C       they must exchange data with each other later. 
        if (i .eq. 1) then
          left=NONE
        else  
          left=i-1
        end if
        if (i .eq. numworkers) then
          right=NONE
        else
          right=i+1
        end if 
C       Now send startup information to each worker
        dest = i
      call MPI_SEND( offset, 1, MPI_INTEGER, dest, BEGIN, 
     &               MPI_COMM_WORLD, ierr )
      call MPI_SEND( cols, 1, MPI_INTEGER, dest, BEGIN, 
     &               MPI_COMM_WORLD, ierr )
      call MPI_SEND( left, 1, MPI_INTEGER, dest, BEGIN, 
     &               MPI_COMM_WORLD, ierr )
      call MPI_SEND( right, 1, MPI_INTEGER, dest, BEGIN, 
     &               MPI_COMM_WORLD, ierr )
      call MPI_SEND( u(1,offset,1), cols * NXPROB, MPI_REAL, dest,
     &               BEGIN, MPI_COMM_WORLD, ierr )
        print *,'Sent to=',dest,'offset=',offset,'cols=',cols,
     .    'left=',left,'right=',right
        offset = offset + cols
      end do
          
C     Now wait for results from all workers tasks 
      do i=1, numworkers
        source = i
        msgtype = DONE
        call MPI_RECV( offset, 1, MPI_INTEGER, source,
     &                 msgtype, MPI_COMM_WORLD, status, ierr )
        call MPI_RECV( cols, 1, MPI_INTEGER, source,
     &                 msgtype, MPI_COMM_WORLD, status, ierr )
        call MPI_RECV( u(1,offset,1), cols * NXPROB, MPI_REAL,
     &                 source,msgtype,MPI_COMM_WORLD,status,ierr)
      end do
 
C     Print and show results 
      print *,'Creating final.dat file and generating graph...'
      call prtdat(2)
      call draw_heat()

 
C     End of master code
      call MPI_FINALIZE(ierr)
      end if
 
      if (taskid .ne. MASTER)  then
C     ****************************** worker code ******************************
C     Initialize everything - including the borders - to zero 
      do ix=1,NXPROB
        do iy=1,NYPROB
          do iz=1,2
            u(ix,iy,iz) = 0.0
          end do
        end do
      end do

C     Now receive my offset, cols, neighbors and grid partition from master 
      source = MASTER
      msgtype = BEGIN
      call MPI_RECV( offset, 1, MPI_INTEGER, source,
     &                 msgtype, MPI_COMM_WORLD, status, ierr )
      call MPI_RECV( cols, 1, MPI_INTEGER, source,
     &                 msgtype, MPI_COMM_WORLD, status, ierr )
      call MPI_RECV( left, 1, MPI_INTEGER, source,
     &                 msgtype, MPI_COMM_WORLD, status, ierr )
      call MPI_RECV( right, 1, MPI_INTEGER, source,
     &                 msgtype, MPI_COMM_WORLD, status, ierr )
      call MPI_RECV( u(1,offset,1),cols*NXPROB,MPI_REAL, source,
     &                 msgtype, MPI_COMM_WORLD, status, ierr )
 
C     Determine border elements.  Need to consider first and last columns.
C     Obviously, col 1 can't exchange with col 1-1.  Likewise, the last
C     col can't exchange with last+1. 
      start=offset
      end=offset+cols-1
      if (offset .eq. 1) then
        start=2
      end if
      if (offset + cols .gt. NYPROB) then
        end=end-1
      end if
      print *,'task=',taskid,'  start=',start,'  end=',end 

C     Begin doing STEPS iterations.  Must communicate border cols with
C     neighbors.  If I have the first or last grid col, then I only need to 
C     communicate with one neighbor. 
      iz=1
      do it=1, STEPS
        if (left .ne. NONE) then
          call MPI_SEND( u(1,offset,iz), NXPROB, MPI_REAL, left,
     &                   RTAG, MPI_COMM_WORLD, ierr )

          source = left
          msgtype = LTAG
          call MPI_RECV( u(1,offset-1,iz),NXPROB,MPI_REAL, source,
     &                   msgtype, MPI_COMM_WORLD, status, ierr )
        end if
 
        if (right .ne. NONE) then
          call MPI_SEND(u(1,offset+cols-1,iz),NXPROB,MPI_REAL,
     &                  right,LTAG,MPI_COMM_WORLD,ierr)

          source = right
          msgtype = RTAG
          call MPI_RECV(u(1,offset+cols,iz),NXPROB,MPI_REAL,source,
     &                   msgtype, MPI_COMM_WORLD, status, ierr )
        end if
      
C       Now call update to update the value of grid points
        call update(start,end,u(1,1,iz),u(1,1,3-iz))
        
        iz=3-iz
      end do
    
C     Send my portion of final results back to master 
      call MPI_SEND( offset, 1, MPI_INTEGER, MASTER, DONE, 
     &               MPI_COMM_WORLD, ierr )
      call MPI_SEND( cols, 1, MPI_INTEGER, MASTER, DONE,
     &               MPI_COMM_WORLD, ierr )
      call MPI_SEND( u(1,offset,iz),cols*NXPROB,MPI_REAL,MASTER,
     &               DONE, MPI_COMM_WORLD, ierr )

C     End of worker code
      call MPI_FINALIZE(ierr)
      end if

      end

C****************************************************************************
      subroutine update (start, end, u1, u2)
C****************************************************************************
      include 'mpi_heat2D.h'
      integer start, end, ix, iy
      real*4 u1, u2
      dimension u1(NXPROB,NYPROB),u2(NXPROB,NYPROB)

      do iy=start, end
        do ix=2, NXPROB-1
            u2(ix,iy) = u1(ix,iy)
     .       + CX * ( u1(ix+1,iy) + u1(ix-1,iy) - 2.0 * u1(ix,iy))
     .       + CY * ( u1(ix,iy+1) + u1(ix,iy-1) - 2.0 * u1(ix,iy))
        end do
      end do
	  end

C*****************************************************************************
      subroutine inidat
C*****************************************************************************
      include 'mpi_heat2D.h'
      integer ix,iy

      do ix=0,NXPROB-1
        do iy=0,NYPROB-1
          u(ix+1,iy+1,1) = float(ix*(NXPROB-ix-1) * iy*(NYPROB-iy-1))
        end do
      end do
      end


C**************************************************************************
      subroutine prtdat(i)
C**************************************************************************
      include 'mpi_heat2D.h'
      integer i,ix, iy
      character*11 fname

      if (i .eq. 1) then
        fname ='initial.dat'
      else if (i .eq. 2) then
        fname = 'final.dat'
      end if

      open(21, file=fname, form='formatted')
      do ix=1,NXPROB
        do iy=1,NYPROB
          write(21,'(f8.1,1x,$)')u(ix,iy,1)
        end do
        write(21,'(1x)')
      end do
      close(21)
      end

File: tutorials/mpi/examples/mpi_heat2D.c
/****************************************************************************
 * FILE: mpi_heat2D.c
 * OTHER FILES: draw_heat.c  
 * DESCRIPTIONS:  
 *   HEAT2D Example - Parallelized C Version
 *   This example is based on a simplified two-dimensional heat 
 *   equation domain decomposition.  The initial temperature is computed to be 
 *   high in the middle of the domain and zero at the boundaries.  The 
 *   boundaries are held at zero throughout the simulation.  During the 
 *   time-stepping, an array containing two domains is used; these domains 
 *   alternate between old data and new data.
 *
 *   In this parallelized version, the grid is decomposed by the master
 *   process and then distributed by rows to the worker processes.  At each 
 *   time step, worker processes must exchange border data with neighbors, 
 *   because a grid point's current temperature depends upon it's previous
 *   time step value plus the values of the neighboring grid points.  Upon
 *   completion of all time steps, the worker processes return their results
 *   to the master process.
 *
 *   Two data files are produced: an initial data set and a final data set.
 *   An X graphic of these two states displays after all calculations have
 *   completed.
 * AUTHOR: Blaise Barney - adapted from D. Turner's serial C version. Converted
 *   to MPI: George L. Gusciora (1/95)
 * LAST REVISED: 06/12/13 Blaise Barney
 ****************************************************************************/
#include "mpi.h"
#include <stdio.h>
#include <stdlib.h>
extern void draw_heat(int nx, int ny);       /* X routine to create graph */

#define NXPROB      20                 /* x dimension of problem grid */
#define NYPROB      20                 /* y dimension of problem grid */
#define STEPS       100                /* number of time steps */
#define MAXWORKER   8                  /* maximum number of worker tasks */
#define MINWORKER   3                  /* minimum number of worker tasks */
#define BEGIN       1                  /* message tag */
#define LTAG        2                  /* message tag */
#define RTAG        3                  /* message tag */
#define NONE        0                  /* indicates no neighbor */
#define DONE        4                  /* message tag */
#define MASTER      0                  /* taskid of first process */

struct Parms { 
  float cx;
  float cy;
} parms = {0.1, 0.1};

int main (int argc, char *argv[])
{
void inidat(), prtdat(), update();
float  u[2][NXPROB][NYPROB];        /* array for grid */
int	taskid,                     /* this task's unique id */
	numworkers,                 /* number of worker processes */
	numtasks,                   /* number of tasks */
	averow,rows,offset,extra,   /* for sending rows of data */
	dest, source,               /* to - from for message send-receive */
	left,right,        /* neighbor tasks */
	msgtype,                    /* for message types */
	rc,start,end,               /* misc */
	i,ix,iy,iz,it;              /* loop variables */
MPI_Status status;


/* First, find out my taskid and how many tasks are running */
   MPI_Init(&argc,&argv);
   MPI_Comm_size(MPI_COMM_WORLD,&numtasks);
   MPI_Comm_rank(MPI_COMM_WORLD,&taskid);
   numworkers = numtasks-1;

   if (taskid == MASTER) {
      /************************* master code *******************************/
      /* Check if numworkers is within range - quit if not */
      if ((numworkers > MAXWORKER) || (numworkers < MINWORKER)) {
         printf("ERROR: the number of tasks must be between %d and %d.\n",
                 MINWORKER+1,MAXWORKER+1);
         printf("Quitting...\n");
         MPI_Abort(MPI_COMM_WORLD, rc);
         exit(1);
         }
      printf ("Starting mpi_heat2D with %d worker tasks.\n", numworkers);

      /* Initialize grid */
      printf("Grid size: X= %d  Y= %d  Time steps= %d\n",NXPROB,NYPROB,STEPS);
      printf("Initializing grid and writing initial.dat file...\n");
      inidat(NXPROB, NYPROB, u);
      prtdat(NXPROB, NYPROB, u, "initial.dat");

      /* Distribute work to workers.  Must first figure out how many rows to */
      /* send and what to do with extra rows.  */
      averow = NXPROB/numworkers;
      extra = NXPROB%numworkers;
      offset = 0;
      for (i=1; i<=numworkers; i++)
      {
         rows = (i <= extra) ? averow+1 : averow; 
         /* Tell each worker who its neighbors are, since they must exchange */
         /* data with each other. */  
         if (i == 1) 
            left = NONE;
         else
            left = i - 1;
         if (i == numworkers)
            right = NONE;
         else
            right = i + 1;
         /*  Now send startup information to each worker  */
         dest = i;
         MPI_Send(&offset, 1, MPI_INT, dest, BEGIN, MPI_COMM_WORLD);
         MPI_Send(&rows, 1, MPI_INT, dest, BEGIN, MPI_COMM_WORLD);
         MPI_Send(&left, 1, MPI_INT, dest, BEGIN, MPI_COMM_WORLD);
         MPI_Send(&right, 1, MPI_INT, dest, BEGIN, MPI_COMM_WORLD);
         MPI_Send(&u[0][offset][0], rows*NYPROB, MPI_FLOAT, dest, BEGIN, 
                  MPI_COMM_WORLD);
         printf("Sent to task %d: rows= %d offset= %d ",dest,rows,offset);
         printf("left= %d right= %d\n",left,right);
         offset = offset + rows;
      }
      /* Now wait for results from all worker tasks */
      for (i=1; i<=numworkers; i++)
      {
         source = i;
         msgtype = DONE;
         MPI_Recv(&offset, 1, MPI_INT, source, msgtype, MPI_COMM_WORLD, 
                  &status);
         MPI_Recv(&rows, 1, MPI_INT, source, msgtype, MPI_COMM_WORLD, &status);
         MPI_Recv(&u[0][offset][0], rows*NYPROB, MPI_FLOAT, source,
                  msgtype, MPI_COMM_WORLD, &status);
      }

      /* Write final output, call X graph and finalize MPI */
      printf("Writing final.dat file and generating graph...\n");
      prtdat(NXPROB, NYPROB, &u[0][0][0], "final.dat");
      printf("Click on MORE button to view initial/final states.\n");
      printf("Click on EXIT button to quit program.\n");
      draw_heat(NXPROB,NYPROB);
      MPI_Finalize();
   }   /* End of master code */



   /************************* workers code **********************************/
   if (taskid != MASTER) 
   {
      /* Initialize everything - including the borders - to zero */
      for (iz=0; iz<2; iz++)
         for (ix=0; ix<NXPROB; ix++) 
            for (iy=0; iy<NYPROB; iy++) 
               u[iz][ix][iy] = 0.0;

      /* Receive my offset, rows, neighbors and grid partition from master */
      source = MASTER;
      msgtype = BEGIN;
      MPI_Recv(&offset, 1, MPI_INT, source, msgtype, MPI_COMM_WORLD, &status);
      MPI_Recv(&rows, 1, MPI_INT, source, msgtype, MPI_COMM_WORLD, &status);
      MPI_Recv(&left, 1, MPI_INT, source, msgtype, MPI_COMM_WORLD, &status);
      MPI_Recv(&right, 1, MPI_INT, source, msgtype, MPI_COMM_WORLD, &status);
      MPI_Recv(&u[0][offset][0], rows*NYPROB, MPI_FLOAT, source, msgtype, 
               MPI_COMM_WORLD, &status);

      /* Determine border elements.  Need to consider first and last columns. */
      /* Obviously, row 0 can't exchange with row 0-1.  Likewise, the last */
      /* row can't exchange with last+1.  */
      start=offset;
      end=offset+rows-1;
      if (offset==0) 
         start=1;
      if ((offset+rows)==NXPROB) 
         end--;
      printf("task=%d  start=%d  end=%d\n",taskid,start,end);

      /* Begin doing STEPS iterations.  Must communicate border rows with */
      /* neighbors.  If I have the first or last grid row, then I only need */
      /*  to  communicate with one neighbor  */
      printf("Task %d received work. Beginning time steps...\n",taskid);
      iz = 0;
      for (it = 1; it <= STEPS; it++)
      {
         if (left != NONE)
         {
            MPI_Send(&u[iz][offset][0], NYPROB, MPI_FLOAT, left,
                     RTAG, MPI_COMM_WORLD);
            source = left;
            msgtype = LTAG;
            MPI_Recv(&u[iz][offset-1][0], NYPROB, MPI_FLOAT, source,
                      msgtype, MPI_COMM_WORLD, &status);
         }
         if (right != NONE)
         {
            MPI_Send(&u[iz][offset+rows-1][0], NYPROB, MPI_FLOAT, right,
                      LTAG, MPI_COMM_WORLD);
            source = right;
            msgtype = RTAG;
            MPI_Recv(&u[iz][offset+rows][0], NYPROB, MPI_FLOAT, source, msgtype,
                      MPI_COMM_WORLD, &status);
         }
         /* Now call update to update the value of grid points */
         update(start,end,NYPROB,&u[iz][0][0],&u[1-iz][0][0]);
         iz = 1 - iz;
      }
      /* Finally, send my portion of final results back to master */
      MPI_Send(&offset, 1, MPI_INT, MASTER, DONE, MPI_COMM_WORLD);
      MPI_Send(&rows, 1, MPI_INT, MASTER, DONE, MPI_COMM_WORLD);
      MPI_Send(&u[iz][offset][0], rows*NYPROB, MPI_FLOAT, MASTER, DONE, 
               MPI_COMM_WORLD);
      MPI_Finalize();
   }
}


/**************************************************************************
 *  subroutine update
 ****************************************************************************/
void update(int start, int end, int ny, float *u1, float *u2)
{
   int ix, iy;
   for (ix = start; ix <= end; ix++) 
      for (iy = 1; iy <= ny-2; iy++) 
         *(u2+ix*ny+iy) = *(u1+ix*ny+iy)  + 
                          parms.cx * (*(u1+(ix+1)*ny+iy) +
                          *(u1+(ix-1)*ny+iy) - 
                          2.0 * *(u1+ix*ny+iy)) +
                          parms.cy * (*(u1+ix*ny+iy+1) +
                         *(u1+ix*ny+iy-1) - 
                          2.0 * *(u1+ix*ny+iy));
}

/*****************************************************************************
 *  subroutine inidat
 *****************************************************************************/
void inidat(int nx, int ny, float *u) {
int ix, iy;

for (ix = 0; ix <= nx-1; ix++) 
  for (iy = 0; iy <= ny-1; iy++)
     *(u+ix*ny+iy) = (float)(ix * (nx - ix - 1) * iy * (ny - iy - 1));
}

/**************************************************************************
 * subroutine prtdat
 **************************************************************************/
void prtdat(int nx, int ny, float *u1, char *fnam) {
int ix, iy;
FILE *fp;

fp = fopen(fnam, "w");
for (iy = ny-1; iy >= 0; iy--) {
  for (ix = 0; ix <= nx-1; ix++) {
    fprintf(fp, "%8.1f", *(u1+ix*ny+iy));
    if (ix != nx-1) 
      fprintf(fp, " ");
    else
      fprintf(fp, "\n");
    }
  }
fclose(fp);
}
